### PR TITLE
refactor: disentangle container config from main worker config in deploy-helpers- #15603 #15603

### PR DESCRIPTION
**Pull request by emily-shen**
Ensures that we do not read container config from the main wrangler config object, but rather from separate container configs (one for regular containers, one for DO-scheduled ones). deploy-helpers now gets Config<Omit, "containers">.

This is necessary to support the new programmatic container config.

---

- Tests
  - [x] Tests included/updated
  - [ ] Automated tests not possible - manual testing has been completed as follows:
  - [ ] Additional testing not necessary because:
- Public documentation
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: refactor

*A picture of a cute animal (not mandatory, but encouraged)*

---

<a href="https://cloudflare.devinenterprise.com/review/cloudflare/workers-sdk/pull/15606" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=3">
    Devin Review
  </picture>
</a>

### CODE REVIEW

File: packages/wrangler/src/deployment-bundle/build-container-images.ts [hunk 1]
```diff
@@ -37,3 +51,104 @@ export async function buildDeployContainerImages(
...
+}
+
+function buildDurableObjectImageTag(
+	scriptName: string,
+	className: string,
+	imageName: string
+): string {
+	const repository = `${scriptName}-${className}-${imageName}`
+		.toLowerCase()
+		.replace(/[^a-z0-9._-]+/g, "-")
```
Comment (dario-piotrowicz) [inline packages/wrangler/src/deployment-bundle/build-container-images.ts:68]: nit: is this replacing `-` with `-`? 🤔

File: packages/wrangler/src/deployment-bundle/build-container-images.ts [hunk 2]
```diff
@@ -37,3 +51,104 @@ export async function buildDeployContainerImages(
...
+		return [];
+	}
+
+	const imagesToBuild = durableObjectContainerConfig.flatMap((container) =>
+		Object.entries(container.images ?? {})
+			.filter((entry): entry is [string, { dockerfile: string }] =>
+				isDockerfileDurableObjectContainerImage(entry[1])
+			)
+			.map(([imageName, image]) => ({ container, imageName, image }))
+	);
```
[the remark this answers is not in the document]
Comment (dario-piotrowicz) [reply packages/wrangler/src/deployment-bundle/build-container-images.ts:101]: @emily-shen could you address this? and include a test to ensure that aliases don't cause duplicate image builds?

**Review by dario-piotrowicz — APPROVED** [review]
Looks good to me, although I think that the [Devin's comment](https://github.com/cloudflare/workers-sdk/pull/15606/changes#r3989420647) should be addressed

File: packages/wrangler/src/deployment-bundle/build-container-images.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (emily-shen) [reply packages/wrangler/src/deployment-bundle/build-container-images.ts:101]: there is no good reason to add the same container image twice, so this is enough of an edgecase that i think this should be ignored for now

[the remark this answers is not in the document]
Comment (emily-shen) [reply packages/wrangler/src/deployment-bundle/build-container-images.ts:101]: e.g. devin's example is ' A Durable Object defines 50 image aliases that all use ./container/Dockerfile.' -> this is so silly

File: packages/wrangler/src/deployment-bundle/build-container-images.ts (hunk 1 above)
Comment (emily-shen) [reply packages/wrangler/src/deployment-bundle/build-container-images.ts:68]: that does look silly but i guess it doesn't hurt 😅

Comment (emily-shen) [reply packages/wrangler/src/deployment-bundle/build-container-images.ts:68]: sorry i am speedrunning this PRs!

**Review by dario-piotrowicz — APPROVED** [review]
Codeowners Bypass

**Comment by workers-devprod** [comment]
Codeowners approval required for this PR:
- @cloudflare/cloudchamber
- ✅ @cloudflare/wrangler

<details><summary>Show detailed file reviewers</summary>

- packages/containers-shared/src/build.ts: [@cloudflare/cloudchamber]
- packages/containers-shared/tests/build-and-push.test.ts: [@cloudflare/cloudchamber]

</details>

**Review by workers-devprod — APPROVED** [review]
Codeowners reviews satisfied
